guppylang/tys/parsing.py

```diff
@@ -13,14 +13,14 @@
 from guppylang.tys.arg import Argument, ConstArg, TypeArg
 from guppylang.tys.const import ConstValue
 from guppylang.tys.param import Parameter, TypeParam
-from guppylang.tys.ty import NoneType, NumericType, TupleType, Type
+from guppylang.tys.ty import InputFlags, NoneType, NumericType, TupleType, Type
 
 
 def arg_from_ast(
     node: AstNode,
     globals: Globals,
     param_var_mapping: dict[str, Parameter] | None = None,
-) -> Argument:
+) -> tuple[Argument, InputFlags]:
```

Comment:
arg! (argh!) Sorry but I got getting really confused here; what is an `Argument`? Is it a value Hugr static TypeArg? (It certainly looks like one) - but then we wouldn't want it to be `inout` or have InputFlags, would we?

I think `arg_from_ast` parses both TypeArgs (Hugr TypeArgs of *kind* TypeParam::Type), and ConstArgs (Hugr TypeArgs of other kinds). But, hang on, ConstArgs are not function parameters (that'd be like `def foo(x: 6)` !). And `@inout` is *only* appropriate for the top/outermost level argument to a `def`, or to `Callable`, yet we make life difficult for *all the other* `check_instantiate`s in order to fit in `Callable` (e.g. `list[int @inout]`) - I see 5 calls to check_no_flags, plus a couple that have no args to have any flags.
* I think passing the flag by subclassing could be an easier route, you *might* still have an equivalent of `check_no_flags` but I reckon it might just fall out of normal `check_all_args` or checking-against-TypeParam (that is: make a normal Hugr TypeParam::Type unable to accept an InoutTypeArg, but Callable has a special TypeParam::MaybeInOutType that can accept both InoutTypeArg and normal TypeArg).
* Alternatively....this'll be even more controversial....can you special-case Callable?

Generally I think I'd favour breaking the lengthy `arg_from_ast` up into `flagged_arg_from_ast` or (`maybe_`)`inout_arg_from_ast` that looks for BinOp with `@` and then (usually) falls back to a shorter `arg_from_ast` that does *not* support that case, and thus, returns a TypeArg (`| ConstArg`) only. (Or returns Argument without separate flags, that'd be a first/lesser step from what's in the PR now).

Which is to say that (my gut feeling is) you'd do better to push error checking downwards towards the parsing/leaves, rather than parsing whatever and then raising error messages (like, found function parameter of type 6). Not sure if that would help with the error locations. Then, say, `maybe_inout_arg_from_ast` would (if not BinOp) call `type_arg_from_ast` (that returns TypeArg, and does not support ConstArg), and then there is `const_or_type_arg_from_ast` (used for arguments to parameterized types), that kind of thing.

Note you've already gone this route with `type_from_arg` vs `type_with_flags_from_arg`, so it's not thaaaat radical....

Reply:
Of course we should also think about likely future annotations. `list[MyStruct @frozen]` is plausible, true....

Reply:
> arg! (argh!) Sorry but I got getting really confused here; what is an Argument? Is it a value Hugr static TypeArg? 

Sorry for the confusion, yes `Argument` is Guppy's version of static arguments to types. E.g. the type `array[int, 5]` has a `TypeArg` `int` and a `ConstArg` `5`.

> More comments would really help, maybe also think about renaming Argument->StaticArg, say.

I added #358 to consider the rename

> And `@inout` is only appropriate for the top/outermost level argument to a def, or to Callable, yet we make life difficult for all the other check_instantiates in order to fit in Callable (e.g. `list[int @inout]`) [...] Alternatively....this'll be even more controversial....can you special-case Callable?

After reflecting a bit on this I think that special casing `Callable` is probaby the best solution in the near term. I initially wanted a general solution that can handle future flags like `@frozen` etc., but I agree it adds a lot of complexity and probably isn't worth it for now. I like your subclassing idea though, I'll probably go for that of we want more general flags in the future 👍

Reply:
> Special-casing Callable might be fine in the short-term, tho, and maybe even in the longer term if it turns out we don't need similar specials for any of the others...

Agreed, I think I'll do that for now 👍

Reply:
Another benefit of special casing `Callable` is that we can now properly handle the list syntax `Callable[[Arg1, Arg2], Return]` without the flattening hack